sw/device/tests/i2c_host_clock_stretching_test.c

```diff
@@ -0,0 +1,255 @@
+ * 3. Checks that only the correct / expected IRQ is triggered.
+ * 4. Clears the IRQ at the peripheral.
+ * 5. Completes the IRQ service at PLIC.
+ */
+static status_t external_isr(void) {
+  dif_rv_plic_irq_id_t plic_irq_id;
+  TRY(dif_rv_plic_irq_claim(&plic, kHart, &plic_irq_id));
+
+  top_earlgrey_plic_peripheral_t peripheral = (top_earlgrey_plic_peripheral_t)
+      top_earlgrey_plic_interrupt_for_peripheral[plic_irq_id];
+  TRY_CHECK(peripheral == kTopEarlgreyPlicPeripheralI2c0,
+            "IRQ from incorrect peripheral: exp = %d(i2c0), found = %d",
+            kTopEarlgreyPlicPeripheralI2c0, peripheral);
+
+  irq_fired =
+      (dif_i2c_irq_t)(plic_irq_id - (dif_rv_plic_irq_id_t)
+                                        kTopEarlgreyPlicIrqIdI2c0FmtThreshold);
+
+  LOG_INFO("%s: plic:%d, i2c:%d", __func__, plic_irq_id, irq_fired);
+  TRY(dif_i2c_irq_acknowledge(&i2c, irq_fired));
+
+  // Complete the IRQ at PLIC.
+  TRY(dif_rv_plic_irq_complete(&plic, kHart, plic_irq_id));
+  return OK_STATUS();
+}
+
+void ottf_external_isr(void) {
+  status_t tmp = external_isr();
+  if (status_ok(isr_result)) {
+    isr_result = tmp;
+  }
+}
+
+// We make this a macro to show the caller line number when printing a error.
+#define CHECK_IRQ_EQ(irq)                                                \
+  do {                                                                   \
+    irq_global_ctrl(false);                                              \
+    if (irq_fired == UINT32_MAX) {                                       \
+      wait_for_interrupt();                                              \
+      irq_global_ctrl(true);                                             \
+    }                                                                    \
+    TRY_CHECK(irq_fired == irq, "Got %d, expected %d.", irq_fired, irq); \
+  } while (false)
+
+static status_t read_manufacture_id(void) {
+  uint8_t reg = kManufacturerIdReg, data = 0;
+  TRY(i2c_testutils_write(&i2c, kDeviceAddr, 1, &reg, true));
+  TRY(i2c_testutils_read(&i2c, kDeviceAddr, 1, &data));
+  TRY_CHECK(data == kManufacturerId, "Unexpected value %x", data);
+  return OK_STATUS();
+}
+
+static status_t read_product_id(void) {
+  uint8_t reg = kProductIdReg, data = 0;
+  TRY(i2c_testutils_write(&i2c, kDeviceAddr, 1, &reg, true));
+  TRY(i2c_testutils_read(&i2c, kDeviceAddr, 1, &data));
+  TRY_CHECK(data == kProductId, "Unexpected value %x", data);
+  return OK_STATUS();
+}
+
+static status_t rx_stretch_timeout(void) {
+  enum { kTimeoutMillis = 15 };
+  TRY(dif_i2c_irq_set_enabled(&i2c, kDifI2cIrqStretchTimeout,
+                              kDifToggleEnabled));
+
+  // Set the delay that will make the target to stretch the clock.
+  {
+    uint8_t write_buffer[2] = {kRxDelayReg, kTimeoutMillis};
+    TRY(i2c_testutils_write(&i2c, kDeviceAddr, sizeof(write_buffer),
+                            write_buffer, true));
+  }
+
+  uint32_t cycles = (kTimeoutMillis - 1) * 100;
```

Comment:
Are there constants in //sw/device/lib/arch/device.h that would make this computation more clear? (ie: something involving a constant with a name like "peripheral clock domain" ?)

Reply:
Done